### PR TITLE
Properly treat utf8-non-encodable yul literals.

### PR DESCRIPTION
**Pull request by chriseth**
References https://github.com/ethereum/solidity/issues/11324

### CODE REVIEW

**Comment by chriseth** [comment]
Should be fixed now.

File: libyul/AsmJsonImporter.cpp [hunk 1]
```diff
@@ -157,7 +157,12 @@ Literal AsmJsonImporter::createLiteral(Json::Value const& _node)
 	auto lit = createAsmNode<Literal>(_node);
 	string kind = member(_node, "kind").asString();
 
-	lit.value = YulString{member(_node, "value").asString()};
+	solAssert(member(_node, "hexValue").isString() || member(_node, "value").isString(), "");
+	if (_node.isMember("hexValue"))
+		lit.value = YulString{util::asString(util::fromHex(member(_node, "hexValue").asString()))};
```
Comment (aarlt) [inline libyul/AsmJsonImporter.cpp:163]: Wouldn't it make sense to also use `util::validateUTF8` here?

Comment (chriseth) [reply libyul/AsmJsonImporter.cpp:163]: No, why? Literals can be invalid utf8

File: Changelog.md [hunk 2]
```diff
@@ -10,6 +10,11 @@ Compiler Features:
 
 
 Bugfixes:
+ * AST: Do not output value of Yul literal if it is not valid utf8.
```
Comment (hrkrshnn) [inline Changelog.md]:
```suggestion
 * AST: Do not output value of Yul literal if it is not a valid utf8 string.
```

File: libyul/AsmJsonImporter.cpp [hunk 3]
```diff
@@ -157,7 +157,12 @@ Literal AsmJsonImporter::createLiteral(Json::Value const& _node)
 	auto lit = createAsmNode<Literal>(_node);
 	string kind = member(_node, "kind").asString();
 
-	lit.value = YulString{member(_node, "value").asString()};
+	solAssert(member(_node, "hexValue").isString() || member(_node, "value").isString(), "");
+	if (_node.isMember("hexValue"))
```
Comment (hrkrshnn) [inline libyul/AsmJsonImporter.cpp:162]: All literal members would now have `hexValue` member, wouldn't it? Why is this in an `if` branch?

Comment (chriseth) [reply libyul/AsmJsonImporter.cpp:162]: True!

Comment (chriseth) [reply libyul/AsmJsonImporter.cpp:162]: I mean this provides some kind of backwards-compatibility, so maybe why not?

Comment (hrkrshnn) [reply libyul/AsmJsonImporter.cpp:162]: Okay. Then just rebasing and the suggestion on changelog should make this ready to merge.

**Comment by chriseth** [comment]
Rebased and updated changelog.

File: test/libsolidity/ASTJSON/assembly/var_access_parseOnly.json [hunk 4]
```diff
@@ -68,6 +68,7 @@
                       "src": "79:6:1",
                       "value":
                       {
+                        "hexValue": "37",
```
Comment (hrkrshnn) [inline test/libsolidity/ASTJSON/assembly/var_access_parseOnly.json]: Value is `7`, but hexValue is `37`?

File: libyul/AsmJsonConverter.cpp [hunk 5]
```diff
@@ -66,7 +67,11 @@ Json::Value AsmJsonConverter::operator()(Literal const& _node) const
 		break;
 	}
 	ret["type"] = _node.type.str();
-	ret["value"] = _node.value.str();
+	if (util::validateUTF8(_node.value.str()))
+		ret["value"] = _node.value.str();
+	else
+		ret["value"] = Json::nullValue;
+	ret["hexValue"] = util::toHex(util::asBytes(_node.value.str()));
```
Comment (hrkrshnn) [inline libyul/AsmJsonConverter.cpp]: If it's an integer literal, should the hex value go through the byte-string conversion?

Comment (chriseth) [reply libyul/AsmJsonConverter.cpp]: Oh you are right, we should only have `hexValue` for string literals.

File: Changelog.md [hunk 6]
```diff
@@ -10,6 +10,11 @@ Compiler Features:
 
 
 Bugfixes:
+ * AST: Do not output value of Yul literal if it is not a valid UTF-8 string.
+
+
+AST Changes:
+ * Add member `hexValue` for Yul literals.
```
Comment (hrkrshnn) [inline Changelog.md]:
```suggestion
 * Add member `hexValue` for Yul string and hex literals.
```

Or maybe just string?

Comment (hrkrshnn) [reply Changelog.md]: Actually, we cannot define a non-utf8 string literal in yul, can we?

Comment (chriseth) [reply Changelog.md]: I'm not sure, but I don't think we need to be that precise here.

**Review by hrkrshnn — APPROVED** [review]

**Pull request merged**